phoenix-core/src/main/java/org/apache/phoenix/mapreduce/VerifyReplicationTool.java

```diff
@@ -0,0 +1,477 @@
+            new Option("tz", "target-zookeeper", true,
+                    "Target ZooKeeper connection details (optional)");
+    static final Option
+            CONDITIONS_OPT =
+            new Option("c", "conditions", true,
+                    "Conditions for select query WHERE clause (optional)");
+    static final Option TIMESTAMP =
+            new Option("ts", "timestamp", true,
+                    "Timestamp in millis used to compare the two tables.  Defaults to current time minus 60 seconds");
+
+    static final Option HELP_OPT = new Option("h", "help", false, "Show this help and quit");
+
+    private Configuration conf;
+
+    private String zkQuorum;
+    private String tableName;
+    private String targetTableName;
+    private String targetZkQuorum;
+    private String sqlConditions;
+    private long timestamp;
+
+    VerifyReplicationTool(Configuration conf) {
+        this.conf = Preconditions.checkNotNull(conf, "Configuration cannot be null");
+    }
+
+    public static Builder newBuilder(Configuration conf) {
+        return new Builder(conf);
+    }
+
+    public static class Verifier
+            extends Mapper<NullWritable, VerifyReplicationSourceWritable, NullWritable, NullWritable> {
+
+        private QueryPlan targetQueryPlan;
+        private PhoenixResultSet targetResultSet = null;
+        private boolean targetHasNext;
+        private boolean sourceHasData;
+
+        public enum Counter {
+            GOODROWS, BADROWS, ONLY_IN_SOURCE_TABLE_ROWS, ONLY_IN_TARGET_TABLE_ROWS,
+            CONTENT_DIFFERENT_ROWS
+        }
+
+        @Override
+        protected void setup(Context context) throws IOException, InterruptedException {
+            super.setup(context);
+            try {
+                targetQueryPlan = PhoenixMapReduceUtil.getQueryPlan(context.getConfiguration(), true);
+            } catch (SQLException e) {
+                throw new IOException(e.getMessage());
+            }
+
+        }
+
+        @Override
+        protected void map(NullWritable key, VerifyReplicationSourceWritable value, Context context)
+                throws IOException, InterruptedException {
+
+            sourceHasData = true;
+            try {
+                if(targetResultSet == null) {
+                    Configuration conf = context.getConfiguration();
+                    byte[] targetStartRow = null;
+                    byte[] targetStopRow = null;
+                    // find source table split
+                    PhoenixInputSplit sourceInputSplit = (PhoenixInputSplit) context.getInputSplit();
+                    if(key != null) {
+                        targetStartRow = value.getSourceKey().get();
+                    }
+                    if(sourceInputSplit.getLength() != 0) {
+                        targetStopRow = sourceInputSplit.getKeyRange().getUpperRange();
+                    }
+                    getTargetResultSet(conf, targetStartRow, targetStopRow);
```

Comment:
Does each map task issue a scan over the target table for with the range of rows withing the map task?

Reply:
@twdsilva yes, for every map task the corresponding target scan is fetched. 
> didn't get the latter part of your statement 'target table for with the range of rows withing the map task'. Can you elaborate?